packages/@aws-cdk/aws-glue-alpha/lib/jobs/job.ts

```diff
@@ -0,0 +1,559 @@
+   *
+   * This metric is based on the Rule returned by no-args onSuccess() call.
+   */
+  public metricSuccess(props?: cloudwatch.MetricOptions): cloudwatch.Metric {
+    return metricRule(this.metricJobStateRule('SuccessMetricRule', JobState.SUCCEEDED), props);
+  }
+
+  /**
+   * Return a CloudWatch Metric indicating job failure.
+   *
+   * This metric is based on the Rule returned by no-args onFailure() call.
+   */
+  public metricFailure(props?: cloudwatch.MetricOptions): cloudwatch.Metric {
+    return metricRule(this.metricJobStateRule('FailureMetricRule', JobState.FAILED), props);
+  }
+
+  /**
+   * Return a CloudWatch Metric indicating job timeout.
+   *
+   * This metric is based on the Rule returned by no-args onTimeout() call.
+   */
+  public metricTimeout(props?: cloudwatch.MetricOptions): cloudwatch.Metric {
+    return metricRule(this.metricJobStateRule('TimeoutMetricRule', JobState.TIMEOUT), props);
+  }
+
+  /**
+   * Creates or retrieves a singleton event rule for the input job state for use with the metric JobState methods.
+   *
+   * @param id construct id.
+   * @param jobState the job state.
+   * @private
+   */
+  private metricJobStateRule(id: string, jobState: JobState): events.Rule {
+    return this.node.tryFindChild(id) as events.Rule ?? this.onStateChange(id, jobState);
+  }
+
+  /**
+   * Returns the job arn
+   * @param scope
+   * @param jobName
+   */
+  protected buildJobArn(scope: constructs.Construct, jobName: string) : string {
+    return cdk.Stack.of(scope).formatArn({
+      service: 'glue',
+      resource: 'job',
+      resourceName: jobName,
+    });
+  }
+}
+
+/**
+ * A subset of Job attributes are required for importing an existing job
+ * into a CDK project. This is ionly used when using fromJobAttributes
+ * to identify and reference the existing job.
+ */
+export interface JobImportAttributes {
+  /**
+   * The name of the job.
+   */
+  readonly jobName: string;
+
+  /**
+   * The IAM role assumed by Glue to run this job.
+   *
+   * @default - undefined
+   */
+  readonly role?: iam.IRole;
+
+}
+
+/**
+ * JobProperties will be used to create new Glue Jobs using this L2 Construct.
+ */
+export interface JobProperties {
+
+  /**
+     * Script Code Location (required)
+     * Script to run when the Glue job executes. Can be uploaded
+     * from the local directory structure using fromAsset
+     * or referenced via S3 location using fromBucket
+     **/
+  readonly script: Code;
+
+  /**
+   * IAM Role (required)
+   * IAM Role to use for Glue job execution
+   * Must be specified by the developer because the L2 doesn't have visibility
+   * into the actions the script(s) takes during the job execution
+   * The role must trust the Glue service principal (glue.amazonaws.com)
+   * and be granted sufficient permissions.
+   *
+   * @see https://docs.aws.amazon.com/glue/latest/dg/getting-started-access.html
+   **/
+  readonly role: iam.IRole;
```

Comment:
How hard can generating a least-privilege IAM Role that has the right execution privileges for a Job script that the CDK construct has no way of knowing it needs? Pretty hard, in fact, nearly impossible, unless the script is exceedingly simple. 

The reason we've decided to require it is because the developer is the only one who knows the permissions the job execution role needs. If we auto-generate a role, we could set the policy to have broad permissions to prevent errors during execution, but this is against AWS best practice. Alternatively, we could set the policy to have very few permissions, which would allow the developer to quickly create the job, but would result in likely execution errors, which the developer would have to update to add the permissions the job script needs. 

When we require it, we provide a better developer experience because we set the expectation up front that the developer needs to know and set the least-privilege permissions the job needs to successfully execute, because they would have to provide that context anyway.

Reply:
I see, thanks for the clarification. I agree with the current approach based on your summary.